R/transform.R

```diff
@@ -0,0 +1,197 @@
+##' Build a parameter transformer, which can be used in models to
+##' translate between an unstructured vector of numbers (the vector
+##' being updated by an MCMC for example) to a structured list of
+##' named values, which is easier to program against.
+##'
+##' @title Build a parameter transformer
+##'
+##' @param scalar Names of scalar parameters.  This is similar for
+##'   listing elements in `array` with values of 1, though elements in
+##'   `scalar` will be placed ahead of those listed in `array` within
+##'   the final parameter vector, and elements in `array` will have
+##'   generated names that include square brackets.
+##'
+##' @param array A list, where names correspond to the names of array
+##'   parameters and values correspond to the lengths of parameters.
+##'   Multiple dimensions are allowed (so if you provide an element
+##'   with two entries these represent dimensions of a matrix).  In
+##'   future, you may be able to use *strings* as values for the
+##'   lengths, in which case these will be looked for within `fixed`.
+##'
+##' @param fixed A named list of fixed parameters; these will be added
+##'   into the final list directly.  These typically represent
+##'   additional pieces of data that your model needs to run, but for
+##'   which you are not performing inference on.
+##'
+##' @param process An arbitrary R function that will be passed the
+##'   final assembled parameter list and will process it - this
+##'   function should return a named list too, but can do whatever
+##'   processing you need.  The default is not to have any processing
+##'   (equivalent to using [identity]).  If you use this, you should
+##'   take care to only append onto your list, and not to modify any
+##'   values - otherwise you will break the generated `untransform`
+##'   method.  We will likely play around with this process in future
+##'   in order to get automatic differentiation to work.
+##'
+##' @return An object of class `mcstate_transformer`, which has three
+##'   elements:
+##'
+##' * `parameters`: a character vector of computed parameter names;
+##'   these are the names that your statistical model will use.
+##' * `transform`: a function that can transform from your statistical
+##'   parameters into a structured list of parameters.
+##' * `untransform`: a function that can untransform from your
+##'   structured list of parameters back into a numeric vector
+##'   suitable for the statistical model.  This cannot untransform
+##'   through a `preprocess` function.
```

Comment:
What does that last sentence mean?

While I'm here - I think the idea of flatten/unflatten is nice here as there's a direction between structured and unstructured (or the reverse) - which I don't feel is quite captured by "un/transform" which feels more generic.

transform here is unflattening (adding the stucture to the flat array)
untransform is flattening (removing the structure and returning a flat array)